---
title: "CROP-seq CRISPRi analysis (Seurat + Mixscape)"
output: html_notebook
---
#load library
```{r}
library(Seurat)
library(ggplot2)
library(patchwork)
library(scales)
library(dplyr)
library(reshape2)
library(Rtsne)
library(readxl)

# Setup custom theme for plotting.
custom_theme <- theme(
  plot.title = element_text(size=16, hjust = 0.5), 
  legend.key.size = unit(0.7, "cm"), 
  legend.text = element_text(size = 14))
```


# 1. Setup a Seurat object
```{r}
# Load the dataset
multimodal.matrix <- Read10X_h5(filename = 'filtered_feature_bc_matrix.h5')
protospacer_calls_per_cell <- read.csv("protospacer_calls_per_cell.csv")

# Creates a Seurat object using all the cells
seurat.obj <- CreateSeuratObject(counts = multimodal.matrix[["Gene Expression"]], min.cells = 0, min.features = 0)
seurat.obj[["CRISPR_Guide"]] <- CreateAssayObject(multimodal.matrix[["CRISPR Guide Capture"]][, colnames(x = seurat.obj)])
seurat.obj[["percent.mt"]] <- PercentageFeatureSet(seurat.obj, pattern = "^MT-")

# Align sgRNA to cells using the CellRanger alignment output file
for(i in 1:nrow(seurat.obj@meta.data)){
  temp <- agrep(rownames(seurat.obj@meta.data)[i],protospacer_calls_per_cell$cell_barcode,max.distance = 0)
  if(length(temp)!=0){
    seurat.obj@meta.data$guide_ID[i] <- protospacer_calls_per_cell$feature_call[temp]
    seurat.obj@meta.data$num_features[i] <- protospacer_calls_per_cell$num_features[temp]
  }else{
    seurat.obj@meta.data$guide_ID[i] <- "Negative"
    seurat.obj@meta.data$num_features[i] <- 0
  }
}

# Visualization for gene expression quality
VlnPlot(seurat.obj, features = c("nFeature_RNA", "nCount_RNA", "percent.mt"), ncol = 3)
```

## Gene expression QC
```{r}
# Remove potential dead or doublet cells
seurat.obj <- subset(seurat.obj, subset=nFeature_RNA<7000 & nFeature_RNA>400 & nCount_RNA< 60000 & percent.mt<10)
seurat.obj

# Visualization for gene expression quality again
FeatureScatter(seurat.obj, feature1 = "nCount_RNA", feature2 = "nFeature_RNA") +
  geom_smooth(method = lm)
FeatureScatter(seurat.obj, feature1 = "nCount_RNA", feature2 = "percent.mt") +
  geom_smooth(method = lm)
```

## Only keep cells with 1 sgRNA for further analysis
```{r}
seurat.obj <- subset(seurat.obj,subset = num_features ==1)
seurat.obj
VlnPlot(seurat.obj, features = c("nFeature_RNA", "nCount_RNA", "percent.mt"), ncol = 3)
```


# 2. RNA-based clustering (Normal single-cell RNA-seq analysis)
```{r}
# find variable features and scale data.
seurat.obj <- NormalizeData(object = seurat.obj) %>% FindVariableFeatures() %>% ScaleData()

# Identify the 10 most highly variable genes
LabelPoints(plot = VariableFeaturePlot(seurat.obj), points = head(VariableFeatures(seurat.obj), 10), repel = TRUE)

# Load the cell cycle markers list (G2/M phase + S phase) from Seurat
s.genes <- cc.genes$s.genes
g2m.genes <- cc.genes$g2m.genes

# Assign Cell-Cycle Scores
seurat.obj <- CellCycleScoring(seurat.obj, s.features = s.genes, g2m.features = g2m.genes, set.ident = TRUE)
View(seurat.obj@meta.data)

# Visualize the distribution of cell cycle markers across
RidgePlot(seurat.obj, features = c("PCNA", "TOP2A", "MCM6", "MKI67"), ncol = 2)

# Run Principle Component Analysis (PCA).
seurat.obj <- RunPCA(object = seurat.obj)
DimHeatmap(seurat.obj, dims = 1:6, cells = 500, balanced = TRUE)
ElbowPlot(seurat.obj)

# Find Clusters
seurat.obj <- FindNeighbors(seurat.obj, dims = 1:15) %>% FindClusters()
seurat.obj <- RunUMAP(object = seurat.obj, dims = 1:15)
DimPlot(seurat.obj, reduction = "umap",label = TRUE)

# Generate plots to check if clustering is driven by cell cycle phase.
DimPlot(
  object = seurat.obj, 
  group.by = 'Phase', 
  label = F, pt.size = 0.2, 
  reduction = "umap", repel = T) + 
  ggtitle("Cell Cycle Phase") +
  ylab("UMAP 2") +
  xlab("UMAP 1") +
  custom_theme
```


# 3.Assign more cell identitiy information to cells (needed in Mixscape)
```{r}
gRNA <- read_excel("gRNA.xlsx")

# Align the "gene" and "crispr" situation into the seurat.obj
for (i in 1:nrow(seurat.obj@meta.data)){
  temp <- agrep(seurat.obj@meta.data$guide_ID[i], gRNA$guide_ID, max.distance = 0)
  if (length(temp) != 0){
    seurat.obj@meta.data$gene[i] <- gRNA$gene[temp]
    seurat.obj@meta.data$crispr[i] <- gRNA$crispr[temp]
    seurat.obj@meta.data$NT[i] <- gRNA$NT[temp]
  }
}
```


# 4. Fold Change calculation
```{r}
# gene level
Idents(seurat.obj) <- "gene"

genes <- unique(seurat.obj[["gene"]][,1])
genes_nn <- genes[!genes == "NT"]

FC_gene <- c()
for(gene in genes_nn){
  temp <- FindMarkers(seurat.obj,ident.1 = gene, ident.2 = "NT", features = gene, pseudocount.use = 1e-9, logfc.threshold=1e-9, min.pct = 1e-9)
  FC_gene <- rbind(FC_gene,temp)
}

write.csv(FC_gene,file = "FC_gene.csv")

# gRNA level
gRNA <- read_excel("gRNA.xlsx")
Idents(seurat.obj) <- "NT"

guides <- sort(unique(seurat.obj[["NT"]][,1]))
guides_nn <- guides[!guides == "NT"]

FC_guide <- c()
for(guide in guides_nn){
  i <- agrep(guide,gRNA$guide_ID,max.distance = 0)
  temp <- FindMarkers(seurat.obj,ident.1 = guide, ident.2 = "NT", features = gRNA$gene[i], pseudocount.use = 1e-9, logfc.threshold=1e-9, min.pct = 1e-9)
  rownames(temp) <- guide
  FC_guide <- rbind(FC_guide,temp)
}

write.csv(FC_guide,file = "FC_guide.csv")
```


# 5. Mixscape - Calculating local perturbation signatures mitigates confounding effects
```{r}
# Calculate perturbation signature (PRTB).
seurat.obj <- CalcPerturbSig(
  object = seurat.obj, 
  assay = "RNA", 
  slot = "data", 
  gd.class ="gene", 
  nt.cell.class = "NT", 
  reduction = "pca", 
  ndims = 40, 
  num.neighbors = 20,            #If have replicate, add: split.by = "replicate",
  new.assay.name = "PRTB")

# Prepare PRTB assay for dimensionality reduction: 
# Normalize data, find variable features and center data.
DefaultAssay(object = seurat.obj) <- 'PRTB'

# Use variable features from RNA assay.
VariableFeatures(object = seurat.obj) <- VariableFeatures(object = seurat.obj[["RNA"]])
seurat.obj <- ScaleData(object = seurat.obj, do.scale = F, do.center = T)

# Run PCA to reduce the dimensionality of the data.
seurat.obj <- RunPCA(object = seurat.obj, reduction.key = 'prtbpca', reduction.name = 'prtbpca')

# Run UMAP to visualize clustering in 2-D.
seurat.obj <- RunUMAP(
  object = seurat.obj, 
  dims = 1:40, 
  reduction = 'prtbpca', 
  reduction.key = 'prtbumap', 
  reduction.name = 'prtbumap')

# Generate plots to check if clustering is driven by cell cycle phase or target gene class
DimPlot(
  object = seurat.obj, 
  group.by = 'Phase', 
  reduction = 'prtbumap', 
  pt.size = 0.2, label = F, repel = T) +
  ggtitle("Cell Cycle Phase") +
  ylab("UMAP 2") +
  xlab("UMAP 1") + 
  custom_theme

DimPlot(
  object = seurat.obj,
  group.by = 'crispr',
  reduction = 'prtbumap', 
  split.by = "crispr", 
  ncol = 1, 
  pt.size = 0.2, 
  cols = c("grey39","goldenrod3")) +
  ggtitle("Perturbation Status") +
  ylab("UMAP 2") +
  xlab("UMAP 1") +
  custom_theme
```


# 6. Mixscape - Identifies cells with no detectable perturbation
```{r}
DefaultAssay(seurat.obj) <- "RNA"

# Run mixscape on sgRNA level!
seurat.obj <- RunMixscape(
  object = seurat.obj, 
  assay = "PRTB", 
  slot = "scale.data", 
  labels = "gene", 
  nt.class.name = "NT", 
  min.de.genes = 5, 
  iter.num = 10, 
  de.assay = "RNA", 
  verbose = F,
  fine.mode = TRUE,
  fine.mode.labels = "guide_ID",
  prtb.type = "KD")

# Save the mixscape result to a csv file
write.csv(reshape2::melt(table(seurat.obj$mixscape_class.global, seurat.obj$NT)),file = "df.csv")

# Add replicate information
seurat.obj@meta.data$MULTI_classification <- "rep1-tx"
seurat.obj@meta.data$MULTI_classification.global <- "rep1-tx"

# Calculate percentage of KD cells for all target gene classes.
df <- prop.table(table(seurat.obj$mixscape_class.global, seurat.obj$NT),2)
df2 <- reshape2::melt(df)
df2$Var2 <- as.character(df2$Var2)
test <- df2[which(df2$Var1 == "KD"),]
test <- test[order(test$value, decreasing = T),]
new.levels <- test$Var2
df2$Var2 <- factor(df2$Var2, levels = new.levels )
df2$Var1 <- factor(df2$Var1, levels = c("NT", "NP", "KD"))
df2$gene <- sapply(as.character(df2$Var2), function(x) strsplit(x, split = "g")[[1]][1])
df2$guide_number <- sapply(as.character(df2$Var2), 
                           function(x) strsplit(x, split = "g")[[1]][2])
df3 <- df2[-c(which(df2$gene == "NT")),]

p1 <- ggplot(df3, aes(x = guide_number, y = value*100, fill= Var1)) +
  geom_bar(stat= "identity") +
  theme_classic()+
  scale_fill_manual(values = c("grey49", "grey79","coral1")) + 
  ylab("Percentage (%) of cells") +
  xlab("sgRNA")

p1 + theme(axis.text.x = element_text(size = 15, hjust = 1), 
           axis.text.y = element_text(size = 12), 
           axis.title = element_text(size = 13), 
           strip.text = element_text(size=13, face = "bold")) + 
  facet_wrap(vars(gene),ncol = 6, scales = "free") +
  labs(fill = "mixscape_class") +theme(legend.title = element_text(size = 11),
                                       legend.text = element_text(size = 9))
```


# 7. Visulization using LDA UMAP with KD cells and NT cells only
## Remove NP cells
```{r}
DefaultAssay(seurat.obj) <- "RNA"
seurat.obj <- subset(seurat.obj,subset = mixscape_class.global != "NP")
seurat.obj
table(seurat.obj$NT)
```

## Form LDA Umap
```{r}
seurat.obj <- MixscapeLDA(
  object = seurat.obj, 
  assay = "RNA", 
  pc.assay = "PRTB", 
  labels = "gene", 
  nt.label = "NT", 
  npcs = 10, 
  logfc.threshold = 0.25, 
  verbose = F)

# Use LDA results to run UMAP and visualize cells on 2-D. 
# Here, we note that the number of the dimensions to be used is equal to the number of 
# labels minus one (to account for NT cells).
seurat.obj <- RunUMAP(
  object = seurat.obj,
  dims = 1:3,
  reduction = 'lda',
  reduction.key = 'ldaumap',
  reduction.name = 'ldaumap')

# Visualize UMAP clustering results.
Idents(seurat.obj) <- "mixscape_class"

DimPlot(seurat.obj, 
        reduction = "ldaumap", 
        repel = T, 
        label.size = 5, 
        label = T)
```


# 8. Calculate DEGs comparing KD cells to NT cells
## Calculate DEGs on the gRNA level
```{r}
DefaultAssay(seurat.obj) <- "RNA"
Idents(seurat.obj) <- "NT"

guides <- sort(unique(seurat.obj$NT))
guides <- guides[guides != "NT"]

# Use logfc.threshold to 0.25 as this is scRNAseq
DEGs <- c()
for(i in guides){
  temp <- FindMarkers(seurat.obj, ident.1 = i, ident.2 = "NT", logfc.threshold = 0.25,test.use = "DESeq2")
  temp <- subset(temp, subset = p_val_adj < 0.05)
  if(length(row.names(temp))>0){
    temp$gRNA <- i
    temp$DEGs <- rownames(temp)
    DEGs <- rbind(DEGs,temp)
  }
}

# Reorder the columns
DEGs <- DEGs[, c(7, setdiff(1:ncol(DEGs), 7))] 
DEGs <- DEGs[, c(7, setdiff(1:ncol(DEGs), 7))]
rownames(DEGs) <- NULL

write.csv(DEGs,file = "gRNA_DEGs_KD_0.25_DESeq2.csv")
```

## Calculate DEGs on the gene level
```{r}
DefaultAssay(seurat.obj) <- "RNA"
Idents(seurat.obj) <- "gene"

genes <- sort(unique(seurat.obj$gene))
genes <- genes[genes != "NT"]


DEGs <- c()
for(i in genes){
  temp <- FindMarkers(seurat.obj, ident.1 = i, ident.2 = "NT", logfc.threshold = 0.25,test.use = "DESeq2")
  temp <- subset(temp, subset = p_val_adj < 0.05)
  if(length(row.names(temp))>0){
    temp$gene <- i
    temp$DEGs <- rownames(temp)
    DEGs <- rbind(DEGs,temp)
  }
}

# Reorder the columns
DEGs <- DEGs[, c(7, setdiff(1:ncol(DEGs), 7))] 
DEGs <- DEGs[, c(7, setdiff(1:ncol(DEGs), 7))]
rownames(DEGs) <- NULL

write.csv(DEGs,file = "Gene_DEGs_KD_0.25_DESeq2.csv")
```
